### ISSUE TITLE
Please add the ability to copy from the diff window

### ISSUE THREAD

**Issue by firedev** [rowanj/gitx#161]
Apart from missing line-wrap #155 it would be great if I could copy from the diff window.

At the moment not only I can't select text (it highlights the whole line), but it's impossible to copy either.

**Comment by WattsInABox** [comment]
+1

**Comment by dragos** [comment]
I'd love this too

**Comment by sugarmo** [comment]
I want this too. And I always felt that this is a basic function.

**Comment by afh** [comment]
I agree copying from the diff window would be great, but only as long as the functionality to stage individual lines or line ranges is kept.

**Comment by afh** [comment]
Pull Request #191 is my proposal to this problem: a button at the top right of the diff view next to the context slider to toggle between the selection modes stage and copy. What do you think?

### Stage mode

![stage](https://f.cloud.github.com/assets/16507/628879/2758ecdc-d106-11e2-8a06-657cb2cf6f3d.png)

### Copy mode

![copy](https://f.cloud.github.com/assets/16507/628880/28136daa-d106-11e2-8f94-5ab39fbc7b82.png)

**Comment by firedev** [comment]
Actually I need to copy when looking at older commits. To re-apply some code or stuff like that. 

Regarding staging, I don't see why it can't allow both copying and staging at the same time. But anyway, I am worrying about copying code from commit diffs. 

Nick

**Comment by afh** [comment]
@firedev Copying from older commits already works for me (see screenshot below).
Please describe in more detail what you would like to do in order to "copy from the diff window".

![screen shot 2013-06-13 at 14 35 14](https://f.cloud.github.com/assets/16507/648267/c4b6a074-d425-11e2-9155-17b2b8a16023.png)

**Comment by firedev** [comment]
You can't copy from Stage window where you mark lines for commit.  

Sorry for my confusing messages prior, I was sending it form my phone without actually taking a look.  

## 

Nick

**Comment by WattsInABox** [comment]
Also when you copy anywhere, multi-line copying doesn't really seem to be possible.

**Comment by tiennou** [comment]
I'm not sure a Stage-Select toggle is the best way to go, because I expect selection and copy-paste to always work, and hiding it behind a toggle seems counter-intuitive IMHO. Maybe it's possible to detect in which direction the selection is started and choose Stage or Select depending on that (selection moving left-right => Select, selection moving top-bottom => Stage) ?

**Comment by jaywilliams** [comment]
Another option would be to use the line numbers to toggle per-line staging. That way the regular cursor selection could be used solely for copying.

**Comment by rowanj** [comment]
I don't think the interface for copying needs to be a mode; copying a selection should Just Work.  I know the HTML/JS interface is terrible for controlling that, but I think adding modes is going to compound the issue more than solve it.

**Comment by wrtsprt** [comment]
I agree, that modes would make it a lot more complicated than necessary. 
Selecting text is already possible, so the solution can't be that far. For a first step being able to copy linewise, would already be great.

**Comment by davidhq** [comment]
+1

**Comment by zeke** [comment]
+1
